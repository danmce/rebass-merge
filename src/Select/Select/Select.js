import React from 'react';
import PropTypes from 'prop-types';
import Button from '../../Button/Button';
import Icon from '../../Icon/Icon';
import {
  ItemStyles, MenuStyles, SelectContainerStyles, SelectContent, SelectStyles,
} from './Select.styles';

export default class Select extends React.Component {
  constructor(props) {
    super(props);
    this.state = { open: this.props.defaultOpen, content: this.props.placeholder };
    this.toggleList = this.toggleList.bind(this);
  }

  selectItem(content, id) {
    this.setState({
      content,
      open: false,
      selected: id,
    });
  }

  toggleList() {
    this.setState((prevState) => ({
      open: !prevState.open,
    }));
  }

  restartSelect(e) {
    e.stopPropagation();
    this.setState({ content: this.props.placeholder, open: false, selected: '' });
  }

  render() {
    return (
      <SelectContainerStyles {...this.props}>
        <SelectStyles onClick={this.toggleList} {...this.props}>
          <SelectContent>
            {this.state.content}
          </SelectContent>
          {this.state.content !== this.props.placeholder
            ? (
              <Button
                mode="minimal"
                type="error"
                size="s"
                stretched={false}
                icon={<Icon icon="ActionCloseSvg" size="xs" color="#444" />}
                onClick={(e) => {
                  this.restartSelect(e);
                  this.props.onRestart();
                }}
              />
            )
            : ''}
          <Icon icon="ChevronSvg" color="#444" />
        </SelectStyles>
        {this.state.open
          ? (
            <MenuStyles {...this.props}>
              {React.Children.map(this.props.children, (item, i) => {
                const option = item.props.children;
                return (
                  <ItemStyles
                    id={`${i}listItem`}
                    active={this.state.selected}
                    onClick={() => {
                      this.selectItem(option, i);
                      this.props.onSelect();
                    }}
                    {...this.props}>
                    {item}
                  </ItemStyles>
                );
              })}
            </MenuStyles>
          ) : ''}
      </SelectContainerStyles>
    );
  }
}

Select.propTypes = {
  children: PropTypes.node,
  width: PropTypes.oneOf(['stretched', 'auto']),
  defaultOpen: PropTypes.bool,
  placeholder: PropTypes.string,
  onSelect: PropTypes.func,
  onRestart: PropTypes.func,
};

Select.defaultProps = {
  placeholder: 'Select...',
  defaultOpen: false,
  width: 'stretched',
};
