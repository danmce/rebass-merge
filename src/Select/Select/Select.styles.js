import styled from 'react-emotion';
import * as colors from '../../styles/colors.json';
import * as shadow from '../../styles/shadows.json';
import * as typography from '../../styles/typography.json';
import * as size from '../../styles/sizes.json';

const getWidth = (props) => {
  if (props.width === 'stretched') {
    return '100%';
  }

  if (props.width === 'auto') {
    return 'auto';
  }

  return `${props.width}px`;
};

const SelectContainerStyles = styled('div')`
    width: ${(props) => (props.stretched ? '100%' : 'auto')};
    font-family: ${typography.fontFamily};
    font-weight: normal;
    font-size: ${typography.size.l}px;
    line-height: ${size.m * typography.lineHeightMultiplier}px;
    color: ${colors.gray.base};
`;

const SelectStyles = styled('div')`
    display: inline-flex;
    flex: 1 0 auto;
    align-items: center;
    justify-content: space-between;
    width: ${(props) => getWidth(props)};
    min-width: 250px;
    height: ${size.xxl}px;
    margin-bottom: 3px;
    border-bottom: 1px solid ${colors.silver.dark20};

    button {
      margin-left: auto;
    }
    svg {
      min-width: fit-content;
    }
`;

const SelectContent = styled('div')`
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
    cursor: pointer;
`;

const MenuStyles = styled('ul')`
    position: absolute;
    display: flex;
    flex-direction: column;
    left: 0;
    width: ${(props) => getWidth(props)};
    min-width: 250px;
    margin: 1px 0;
    padding: ${size.s}px 0 ${size.s}px 0;
    background: white;
    box-shadow: ${shadow.box};
    list-style: none;
`;

const ItemStyles = styled('li')`
    display: inline-flex;
    flex-direction: row;
    align-items: center;
    justify-content: flex-start;
    position: relative;
    width: inherit;
    height: ${size.xxl}px;
    padding: 0 ${size.m}px;
    background-color: ${(props) => (props.id === `${props.active}listItem` ? colors.silver.light1 : 'white')};
    cursor: pointer;
    &:hover {
        background-color: ${colors.silver.light1};
    }
    span {
        max-width: 90%;
        overflow: hidden;
        white-space: nowrap;
        text-align: center;
        text-overflow: ellipsis;
    }
`;

export {
  SelectStyles, MenuStyles, ItemStyles, SelectContainerStyles, SelectContent,
};
