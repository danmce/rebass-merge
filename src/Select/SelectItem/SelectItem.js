import React from 'react';
import PropTypes from 'prop-types';

export default function SelectItem(props) {
  return (
    <span>
      {props.children}
    </span>
  );
}

SelectItem.propTypes = {
  children: PropTypes.string,
};
