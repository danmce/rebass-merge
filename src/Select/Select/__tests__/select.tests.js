import React from "react";
import * as enzyme from "enzyme";
import toJson from "enzyme-to-json";
import Select from "../Select";
import SelectItem from "../../SelectItem/SelectItem";

test("renders Select with width stretched correctly", () => {
    const select = (
      <Select width="stretched">
        <SelectItem>Option 1</SelectItem>
        <SelectItem>Option 2</SelectItem>
        <SelectItem>Option 3</SelectItem>
        <SelectItem>Option 4</SelectItem>
        <SelectItem>Option 5</SelectItem>
        <SelectItem>Option 5 Option 5 Option 5 Option 5 Option 5</SelectItem>
      </Select>
    );
  
    expect(toJson(enzyme.shallow(select))).toMatchSnapshot(`enzyme.shallow`);
  });

test("renders Select with width auto correctly", () => {
    const select = (
      <Select width="auto">
        <SelectItem>Option 1</SelectItem>
        <SelectItem>Option 2</SelectItem>
        <SelectItem>Option 3</SelectItem>
        <SelectItem>Option 4</SelectItem>
        <SelectItem>Option 5</SelectItem>
        <SelectItem>Option 5 Option 5 Option 5 Option 5 Option 5</SelectItem>
      </Select>
    );
  
    expect(toJson(enzyme.shallow(select))).toMatchSnapshot(`enzyme.shallow`);
  });