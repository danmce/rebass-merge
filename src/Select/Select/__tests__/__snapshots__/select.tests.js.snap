// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Select with width auto correctly: enzyme.shallow 1`] = `
<Styled(div)
  defaultOpen={false}
  placeholder="Select..."
  width="auto"
>
  <Styled(div)
    defaultOpen={false}
    onClick={[Function]}
    placeholder="Select..."
    width="auto"
  >
    <Styled(div)>
      Select...
    </Styled(div)>
    <Icon
      color="#444"
      icon="ChevronSvg"
    />
  </Styled(div)>
</Styled(div)>
`;

exports[`renders Select with width stretched correctly: enzyme.shallow 1`] = `
<Styled(div)
  defaultOpen={false}
  placeholder="Select..."
  width="stretched"
>
  <Styled(div)
    defaultOpen={false}
    onClick={[Function]}
    placeholder="Select..."
    width="stretched"
  >
    <Styled(div)>
      Select...
    </Styled(div)>
    <Icon
      color="#444"
      icon="ChevronSvg"
    />
  </Styled(div)>
</Styled(div)>
`;
